.azure-pipelines/azure-pipelines-build.yml

```diff
@@ -1,72 +1,82 @@
+# The azure pipeline template for Official build, and upgrade version build
+
 parameters:
 - name: 'jobFilters'
   type: object
   default: ''
 - name: 'buildOptions'
   type: string
   default: 'SONIC_CONFIG_BUILD_JOBS=1'
-- name: 'buildSlave'
-  type: string
-  default: 'n'
+- name: 'preSteps'
+  type: stepList
+  default: []
 - name: 'postSteps'
   type: stepList
   default: []
 
 jobs:
-- template: azure-pipelines-job-groups.yml
+- template: azure-pipelines-image-template.yml
   parameters:
     jobFilters: ${{ parameters.jobFilters }}
-    preSteps:
-      - script: |
-          containers=$(docker container ls | grep "sonic-slave" | awk '{ print $1 }')
-          if [ ! -z "$containers" ]; then
-            docker container kill $containers || true
-            sleep 5
-          fi
-          if [ "${{ parameters.buildSlave }}" == "y" ]; then
-            images=$(docker images 'sonic-slave-*' -a -q)
-            [ ! -z "$images" ] && docker rmi -f $images
-          fi
-          sudo rm -rf $(ls -A1)
-        displayName: 'Init'
-      - checkout: self
-        submodules: recursive
-        displayName: 'Checkout code'
-      - script: |
-          make ${{ parameters.buildOptions }} PLATFORM=$GROUP_NAME configure
-        displayName: 'Make configure'
-    postSteps:
-      - ${{ parameters.postSteps }}
-      - publish: $(System.DefaultWorkingDirectory)/target
-        artifact: 'sonic-buildimage.$(GROUP_NAME)$(GROUP_EXTNAME)'
-        displayName: "Archive sonic image"
+    buildOptions: ${{ parameters.buildOptions }}
+    preSteps: ${{ parameters.preSteps }}
+    postSteps: ${{ parameters.postSteps }}
     jobGroups:
       - name: vs
         script: |
-          sudo bash -c "echo 1 > /proc/sys/vm/compact_memory"
-          make ${{ parameters.buildOptions }} target/sonic-vs.img.gz
+          make $(BUILD_OPTIONS) INSTALL_DEBUG_TOOLS=y target/sonic-vs.img.gz
+          mv target/sonic-vs.img.gz target/sonic-vs-dbg.img.gz
+          make $(BUILD_OPTIONS) target/docker-sonic-vs.gz target/sonic-vs.img.gz target/docker-ptf.gz
       - name: broadcom
         script: |
-          make ${{ parameters.buildOptions }} target/sonic-broadcom.bin target/sonic-aboot-broadcom.swi
+          make $(BUILD_OPTIONS) INSTALL_DEBUG_TOOLS=y target/sonic-broadcom.bin
+          mv target/sonic-broadcom.bin target/sonic-broadcom-dbg.bin
+          make $(BUILD_OPTIONS) target/sonic-broadcom.bin 
+          make $(BUILD_OPTIONS) ENABLE_IMAGE_SIGNATURE=y target/sonic-aboot-broadcom.swi
+          make $(BUILD_OPTIONS) target/sonic-broadcom.raw
+          make $(BUILD_OPTIONS) ENABLE_SYNCD_RPC=y target/docker-syncd-brcm-rpc.gz
+      - name: mellanox
+        script: |
+          make $(BUILD_OPTIONS) INSTALL_DEBUG_TOOLS=y target/sonic-mellanox.bin
+          mv target/sonic-mellanox.bin target/sonic-mellanox-dbg.bin
+          make $(BUILD_OPTIONS) ENABLE_SYNCD_RPC=y target/sonic-aboot-mlnx-rpc.gz
```

Comment:
i do not understand the motivation here. it looks going backward. readability is low.

Reply:
@lguohan , it makes the build script more flexible, moving the no platform relative scripts into a template, such as initialization and cleaning up, archiving artifacts. We do not need to handle if/else logic relative to platform in the template.
When we change the script of one platform, it is not necessary to impact the other platform, so easy to satisfy to different requirement in each of the branches, PR and official build. For a simple example, when give debug=true, we may only need debug build only, or bebug and release build, we may like to build release build and then build debug build. And other example, we like to change target name in the script.

Reply:
Another motivation is that, we can separate the official build into multiple azure pipelines definitions or in one azure pipelines definition without changing the yaml file.